### ISSUE TITLE
Problem with transcode

### ISSUE THREAD

**Issue by DOFfactory** [linuxserver/docker-mstream#4]
Running mStream from docker (`linuxserver/mstream` version), with PUID=1000 and PGID=100. Whenever I start the container, I receive the following error in logs:

```
  v4.5.2     ____  _
   _ __ ___ / ___|| |_ _ __ ___  __ _ _ __ ___
  | '_ ` _ \\___ \| __| '__/ _ \/ _` | '_ ` _ \
  | | | | | |___) | |_| | |  __/ (_| | | | | | |
  |_| |_| |_|____/ \__|_|  \___|\__,_|_| |_| |_|
  Paul Sori - paul@mstream.io

Find a bug? Report it at:
https://github.com/IrosTheBeggar/mStream/issues

4
    throw err;
    ^

Error: EACCES: permission denied, mkdir '/root/.ffbinaries-cache'
    at Object.mkdirSync (fs.js:757:3)
    at ensureDirSync (/opt/mstream/node_modules/ffbinaries/ffbinaries-lib.js:26:9)
    at Object.<anonymous> (/opt/mstream/node_modules/ffbinaries/ffbinaries-lib.js:30:1)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)

```
What I am able to identify is that the issue is related with the transcode section in config.json, particularly with the path to ffmpeg. I installed to alpine linux (inside of the docker image) ffmpeg, but mstream still cannot find it. 

I tried with both paths 

```
root@22cfe32a9a12:/# whereis ffmpeg
ffmpeg: /usr/bin/ffmpeg /usr/share/ffmpeg
```

Could you help?

**Comment by DOFfactory** [comment]
Hm, in addition, every time the docker is restarted, the installed `ffmpeg` disappears

**Comment by tobbenb** [comment]
A restart should not remove an installed package.
You can use our custom init system to install packages. On mobile so I don't have the link, but it's on the blog part of our website.

**Comment by DOFfactory** [comment]
Well, I just restarted my `linuxserver/mstream` docker, where I had previously installed ffmpeg, and you can see the commands yourself

```
root@22cfe32a9a12:/# ls /opt/mstream/
.gitignore           README.md            cli-boot-wrapper.js  docs/                ffmpeg/              image-cache/         mstream-electron.js  node_modules/        package.json         save/
LICENSE              build/               dev-app-update.yml   electron/            frp/                 modules/             mstream.js           package-lock.json    public/              sync/
root@22cfe32a9a12:/# ls /opt/mstream/ffmpeg/README.md
.dockerenv   bin/         defaults/    docker-mods  home/        lib/         media/       music/       proc/        run/         srv/         tmp/         var/
app/         config/      dev/         etc/         init         libexec/     mnt/         opt/         root/        sbin/        sys/         usr/
root@22cfe32a9a12:/# ls /opt/mstream/ffmpeg/README.md
.dockerenv   bin/         defaults/    docker-mods  home/        lib/         media/       music/       proc/        run/         srv/         tmp/         var/
app/         config/      dev/         etc/         init         libexec/     mnt/         opt/         root/        sbin/        sys/         usr/
root@22cfe32a9a12:/# apk add ffmpeg
fetch http://dl-cdn.alpinelinux.org/alpine/v3.10/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.10/community/x86_64/APKINDEX.tar.gz
(1/42) Installing sdl2 (2.0.10-r0)
(2/42) Installing libxau (1.0.9-r0)
(3/42) Installing libbsd (0.9.1-r0)
(4/42) Installing libxdmcp (1.1.3-r0)
(5/42) Installing libxcb (1.13.1-r0)
(6/42) Installing libx11 (1.6.8-r1)
(7/42) Installing alsa-lib (1.1.9-r0)
(8/42) Installing expat (2.2.8-r0)
(9/42) Installing libbz2 (1.0.6-r7)
(10/42) Installing libpng (1.6.37-r1)
(11/42) Installing freetype (2.10.0-r0)
(12/42) Installing libuuid (2.33.2-r0)
(13/42) Installing fontconfig (2.13.1-r0)
(14/42) Installing fribidi (1.0.5-r2)
(15/42) Installing libass (0.14.0-r0)
(16/42) Installing gmp (6.1.2-r1)
(17/42) Installing nettle (3.4.1-r1)
(18/42) Installing libffi (3.2.1-r6)
(19/42) Installing p11-kit (0.23.16.1-r0)
(20/42) Installing libtasn1 (4.14-r0)
(21/42) Installing libunistring (0.9.10-r0)
(22/42) Installing gnutls (3.6.8-r0)
(23/42) Installing lame (3.100-r0)
(24/42) Installing opus (1.3.1-r0)
(25/42) Installing libogg (1.3.3-r2)
(26/42) Installing libtheora (1.1.1-r14)
(27/42) Installing libjpeg-turbo (2.0.3-r0)
(28/42) Installing v4l-utils-libs (1.16.6-r0)
(29/42) Installing libxext (1.3.4-r0)
(30/42) Installing libxfixes (5.0.3-r2)
(31/42) Installing libpciaccess (0.14-r0)
(32/42) Installing libdrm (2.4.98-r0)
(33/42) Installing wayland-libs-client (1.17.0-r0)
(34/42) Installing libva (2.4.1-r0)
(35/42) Installing libvdpau (1.2-r0)
(36/42) Installing libvorbis (1.3.6-r2)
(37/42) Installing libvpx (1.8.0-r0)
(38/42) Installing x264-libs (20180304-r1)
(39/42) Installing x265-libs (3.0-r0)
(40/42) Installing xvidcore (1.3.4-r1)
(41/42) Installing ffmpeg-libs (4.1.4-r0)
(42/42) Installing ffmpeg (4.1.4-r0)
Executing busybox-1.30.1-r3.trigger
OK: 126 MiB in 78 packages
root@22cfe32a9a12:/# 
```

Adding the following lines to `config.json` does not help:

```
  "transcode": {
    "enabled": true,
    "ffmpegDirectory": "/path/to/ffmpeg-dir",
    "defaultCodec": "opus",
    "defaultBitrate": "128k"
  }
```

It gives me the errors posted in the 1st post.

**Comment by tobbenb** [comment]
I don't understand what those directory listing are supposed to show? Looks like you are trying to list the Readme file in the two last options.

If you install ffmpeg using apk it will not install in /opt/mstream.
As for the error message I haven't checked if it's happening here. Ffmpeg might be statically for mstream.

**Comment by DOFfactory** [comment]
If you check that Readme.md file, it contains the following message

```
ffmpeg binaries will be downloaded here unless otherwise specified
Use the -f flag to set location manually
```
I use the `linuxserver/mstream:latest` from hub.docker.com. Yes, I am using the json config.

So I was expecting as soon as I add the transcode part into the `config.json` mstream will download the corresponding ffmpeg. It did not happen. Also, if I install ffmpeg with the command `apk add ffmpeg` and then restart the docker (stop-start), the ffmpeg disappears. I know this is unusual behaviour, that is the reason I am reporting it. It should indeed keep the installed app. Or, adjust the docker container build, and add the ffmpeg installation by default (and to the correct location, i.e. `/opt/mstream/ffmpeg`).

**Comment by tobbenb** [comment]
I know it's downloading the ffmpeg and I said so in my previous post, but it's only downloaded if the -f command line switch is set.
It's not enough to add the path to the json. You also need to set a variable when creating the container to use the json.
Post the command you used to start the container or the docker compose you used.

Ffmpeg is installed in /usr/bin/

If ffmpeg is disappearing from /usr/bin after a restart, then you have an issue with your docker install/system. That is not a container issue.

**Comment by DOFfactory** [comment]
Hi, there is nothing wrong with my docker system, as all the other containers can survive a restart with the additionally installed packages. But this is not my issue. I wonder why you are not implementing the transcoding part by default, and eventually the user can disable it by removing the parts in json, instead of being forced to install additional packages and then face issues? Would be easier for both of us ... imho

**Comment by tobbenb** [comment]
You are welcome to do a PR.
If you want help here, please provide the requested info.

**Comment by DOFfactory** [comment]
Here you go. 

docker configs:
```
docker create \
  --name=mstream \
  -e PUID=1000 \
  -e PGID=100 \
  -e USE_JSON=true\
  -e TZ=Europe/London \
  -p 3000:3000 \
  -v /home/docker/mstream/config:/config \
  -v /volume1/Music:/music \
  --restart unless-stopped \
  linuxserver/mstream
```

config.json
```
{
  "port": 3000,
  "webAppDirectory": "public",
  "writeLogs": false,
  "noUpload": true,
  "scanOptions": {
    "skipImg": false,
    "scanInterval": 1.5,
    "pause": 50,
    "saveInterval": 500,
    "bootScanDelay": 15
  },
  "transcode": {
    "enabled": true,
    "defaultCodec": "opus",
    "defaultBitrate": "128k"
  },
  "storage": {
    "albumArtDirectory": "/config/album-art",
    "dbDirectory": "/config/db"
  },
  "folders": {
    "library": { "root": "/music" }
  },
  "users": {
    "admin": {
      "password":"password",
      "vpaths": ["library"]
      }
  }
}
```

logs
```
Error: EACCES: permission denied, mkdir '/root/.ffbinaries-cache',
    at Object.mkdirSync (fs.js:757:3),
    at ensureDirSync (/opt/mstream/node_modules/ffbinaries/ffbinaries-lib.js:26:9),
    at Object.<anonymous> (/opt/mstream/node_modules/ffbinaries/ffbinaries-lib.js:30:1),
    at Module._compile (internal/modules/cjs/loader.js:778:30),
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10),
    at Module.load (internal/modules/cjs/loader.js:653:32),
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12),
    at Function.Module._load (internal/modules/cjs/loader.js:585:3),
    at Module.require (internal/modules/cjs/loader.js:692:17),
    at require (internal/modules/cjs/helpers.js:25:18),
```

As soon as I remove the following part from `config.json`:
```
  "transcode": {
    "enabled": true,
    "defaultCodec": "opus",
    "defaultBitrate": "128k"
  },
```
the docker starts up and runs fine.

Please note that there is some permission issue in the logs. I would like to use this docker, however, **with** transcoding.

**Comment by tobbenb** [comment]
You have to install ffmpeg to get it to work with transcoding. It will not work with the ffmpeg binary downloaded by mstream.

So you need to add the path to the ffmpeg binary in the transcode section of the json.

"ffmpegDirectory": "/usr/bin/ffmpeg",

Ffmpeg should already be installed from your previous apk add ffmpeg, unless you recreated the container. If you did, you have to run the command again, or better, learn how to customize our container to install ffmpeg. 
https://blog.linuxserver.io/2019/09/14/customizing-our-containers/

**Comment by DOFfactory** [comment]
Ok, this is what I did:

1. I log in to the docker, installed ffmpeg. Checked, it is installed successfully.
2. edit the config.json file from within the docker, adding the line that you just listed above
3. checked the logs, and the html page of mstream, transcoding is not recognized yet (it says under Transcoding: "Transcoding is disabled on this server")
4. restarted the docker
5. the permission denied error is back.

I am telling you, something is wrong, either with the priviledges, or the fact that I am using PUID/PGID settings that are somehow not working.
